LibCarla/source/carla/trafficmanager/CollisionStage.cpp

```diff
@@ -72,51 +76,45 @@ namespace CollisionStageConstants {
+          const ActorId other_actor_id = j->first;
+          const cg::Location other_location = other_actor->GetLocation();
+
+          // Collision checks increase with speed
+          float collision_distance = std::pow(floor(ego_actor->GetVelocity().Length()*3.6f/10.0f),2.0f);
+          collision_distance = cg::Math::Clamp(collision_distance, MIN_COLLISION_RADIUS, MAX_COLLISION_RADIUS);
 
-      // Continue only if random number is lower than our %, default is 0.
-      if (parameters.GetPercentageIgnoreActors(boost::shared_ptr<cc::Actor>(ego_actor)) <= r) {
-        // Check every actor in the vicinity if it poses a collision hazard.
-        for (auto j = overlapping_actors.begin(); (j != overlapping_actors.end()) && !collision_hazard; ++j) {
-          try {
-            const Actor actor = j->second;
-            const ActorId actor_id = j->first;
-            const cg::Location other_location = actor->GetLocation();
+          // Temporary fix to (0,0,0) bug
+          if (!(other_location.x == 0 && other_location.y == 0 && other_location.z == 0)) {
 
-            // Collision checks increase with speed (Official formula used)
-            float collision_distance = std::pow(floor(ego_actor->GetVelocity().Length()*3.6f/10.0f),2.0f);
-            collision_distance = cg::Math::Clamp(collision_distance, MIN_COLLISION_RADIUS, MAX_COLLISION_RADIUS);
+            if (other_actor_id != ego_actor_id &&
+                (cg::Math::DistanceSquared(ego_location, other_location)
+                < std::pow(MAX_COLLISION_RADIUS, 2)) &&
+                (std::abs(ego_location.z - other_location.z) < VERTICAL_OVERLAP_THRESHOLD)) {
 
-            // Temporary fix to (0,0,0) bug
-            if (other_location.x != 0 && other_location.y != 0 && other_location.z != 0){
+              if (parameters.GetCollisionDetection(ego_actor, other_actor)) {
 
-              if (actor_id != ego_actor_id &&
-                  (cg::Math::DistanceSquared(ego_location, other_location)
-                  < std::pow(MAX_COLLISION_RADIUS, 2)) &&
-                  (std::abs(ego_location.z - other_location.z) < VERTICAL_OVERLAP_THRESHOLD)) {
+                if((safe_point_junction != nullptr && !IsLocationAfterJunctionSafe(ego_actor, other_actor, safe_point_junction, other_location)) ||
+                  NegotiateCollision(ego_actor, other_actor, ego_location, other_location, closest_point, junction_look_ahead)) {
 
-                if (safe_point_junction != nullptr){
-                  if(parameters.GetCollisionDetection(ego_actor, actor) &&
-                    !IsLocationAfterJunctionSafe(ego_actor, actor, safe_point_junction)){
+                  if ((other_actor_type[0] == 'v' && parameters.GetPercentageIgnoreVehicles(ego_actor) <= (rand() % 101)) ||
+                      (other_actor_type[0] == 'w' && parameters.GetPercentageIgnoreWalkers(ego_actor) <= (rand() % 101))) {
 
                     collision_hazard = true;
                     break;
```

Comment:
This break statement seems unnecessary. The looping condition at line 80 already tracks the state of variable collision_hazard and exits if it's value becomes true.